kubernetes/core/src/main/kotlin/com/google/kubernetes/tools/core/KubectlExecutorService.kt

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.kubernetes.tools.core
+
+import com.intellij.execution.ExecutionException
+import com.google.cloud.tools.intellij.analytics.UsageTrackerService
+import com.intellij.openapi.components.ServiceManager
+import java.io.BufferedReader
+import java.io.InputStreamReader
+import java.util.concurrent.TimeUnit
+
+const val KUBECTL_SUCCESS = "kubectl.run"
+const val KUBECTL_FAIL = "kubectl.fail"
+const val METADATA_ERROR_MESSAGE_KEY = "error.message"
+
+class KubectlExecutorService {
+    companion object {
+        val instance
+            get() = ServiceManager.getService(KubectlExecutorService::class.java)!!
+    }
+
+    var kubectlExecutablePath: String = "kubectl"
+
+    /**
+     * run kubectl version to make sure that kubectl is available, if there's no response after
+     *  a couple of seconds then it is not available.
+     */
+    fun isKubectlAvailable(): Boolean =
+        try {
+            val settings = KubectlExecutorSettings(KubectlExecutorSettings.ExecutionMode.VERSION)
+            val process = executeKubectl(settings).process
+            process.waitFor(2, TimeUnit.SECONDS)
+            process.exitValue() == 0
+        } catch (e: Exception) {
+            false
+        }
+
+    /**
+     * Parse through command line arguments and try to execute the command
+     */
+    fun executeKubectl(settings: KubectlExecutorSettings): KubectlProcess {
+
+        val commandList = mutableListOf<String>()
+
+        with(commandList) {
+            add(kubectlExecutablePath)
+            add(settings.executionMode.modeFlag)
+            settings.executionFlags.forEach {
+                add(it)
+            }
+        }
+
+        try {
+            val processBuilder = ProcessBuilder()
+            val startedProcess = processBuilder.command(commandList).start()
+            val kubectlProcess = KubectlProcess(
+                    startedProcess,
+                    commandList.joinToString(" ")
+            )
+            // track event based on execution mode.
+            UsageTrackerService.getInstance().trackEvent(KUBECTL_SUCCESS).ping()
```

Comment:
Could we leave the success metrics ping out of this PR for now - we probably don't need to know when a user launches kubectl, but rather interested in more high-level actions, i.e. list clusters. etc. We can add them later when we have them. But we can keep failure one to know failures and reasons.